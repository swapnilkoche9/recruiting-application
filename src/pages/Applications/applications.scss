.container {
  display: flex;
  flex-direction: column;
  padding-left: 50px;
  padding-right: 50px;
}
.applications_header {
  margin: '10px';
  display: flex;
}

.input-wrapper {
  display: flex;
  justify-content: flex-end;
  width: 100%;
  margin-right: 50px;
  margin-bottom: 20px;
}

.info-wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
}
