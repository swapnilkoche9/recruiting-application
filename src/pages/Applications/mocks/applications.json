{
  "data": [
    {
      "id": 1,
      "name": "Alvin Satterfield",
      "email": "cornellbartell@connellyleannon.biz",
      "birth_date": "1997-09-07",
      "year_of_experience": 5,
      "position_applied": "Technician",
      "application_date": "2018-07-02",
      "status": "rejected"
    },
    {
      "id": 2,
      "name": "Colette Morar",
      "email": "corinnestark@pacocha.co",
      "birth_date": "1998-08-03",
      "year_of_experience": 3,
      "position_applied": "Designer",
      "application_date": "2017-11-18",
      "status": "waiting"
    }
  ]
}
