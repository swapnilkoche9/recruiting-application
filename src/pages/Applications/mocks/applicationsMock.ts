import { rest } from 'msw'
import applicationsMock from './applications.json'

export enum ApplicationsScenarios {
  ERROR = 'ERROR',
}

export const baseHandlers = () => [
  rest.get('http://personio-fe-test.herokuapp.com/api/v1/candidates', (req, res, ctx) => {
    return res(ctx.json(applicationsMock))
  }),
]

export const handlerForScenario = {
  [ApplicationsScenarios.ERROR]: () =>
    rest.get('http://personio-fe-test.herokuapp.com/api/v1/candidates', (req, res, ctx) => {
      return res(ctx.json({ error: { code: 500, message: 'Fatal error :(' } }))
    }),
}
