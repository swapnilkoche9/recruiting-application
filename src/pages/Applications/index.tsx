import { FC, useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { useQuery } from 'react-query'
import { useLocation } from 'react-router-dom'

import { applicationsAdaptor } from '../../adapters/applicationsAdapter'
import Table from '../../components/Table'
import TableRow from '../../components/Table/Row'
import {
  formattedHashValues,
  formattedRowData,
  setUrlParams,
  sortAndFilterApplications,
} from './utils/util'
import { Column, IApplication } from '../../dtos/applications'
import InfoBox from '../../components/InfoBox'
import InputBox from '../../components/InputBox'
import TableHeader from '../../components/Table/Header'

import './applications.scss'

const columns: Column[] = [
  { name: 'Name', canBeSorted: false, key: 'name' },
  { name: 'Email', canBeSorted: false, key: 'email' },
  { name: 'Age', canBeSorted: false, key: 'birth_date' },
  { name: 'Years of experience', canBeSorted: true, key: 'year_of_experience' },
  { name: 'Position applied', canBeSorted: true, key: 'position_applied' },
  { name: 'Applied', canBeSorted: true, key: 'application_date' },
  { name: 'Status', canBeSorted: false, key: 'status' },
]

export type SortTypes = 'ascending' | 'descending'
export type SortKeys = keyof IApplication

interface ApplicationResponse {
  data?: IApplication[]
  error?: { code: number; message: string }
}

const Applications: FC = () => {
  const [sortValue, setSortValue] = useState<SortKeys>()
  const [searchValue, setSearchValue] = useState('')
  const sortType = useRef<SortTypes>('ascending')

  const { hash } = useLocation()

  const {
    data = { data: [] },
    isLoading,
    isError,
  } = useQuery<ApplicationResponse>('application', applicationsAdaptor)

  const { data: applications = [], error = null } = data
  const errorOccured = !!error || isError

  const {
    sortKey = '',
    activeSort = '',
    searchKey = '',
  } = useMemo(() => formattedHashValues(hash), [hash])

  const formattedSearchKey = useMemo(() => searchKey.split('+').join(' '), [searchKey])

  const filteredApplications = useMemo(
    () =>
      sortAndFilterApplications(
        applications,
        formattedSearchKey ?? searchValue,
        (sortKey as keyof IApplication) ?? sortValue,
        (activeSort as SortTypes) ?? sortType.current,
      ) ?? [],
    [applications, formattedSearchKey, searchValue, sortKey, sortValue, activeSort],
  )

  const showApplications = !errorOccured && !isLoading && filteredApplications.length > 0

  const noResult = filteredApplications.length === 0 && !errorOccured && !isLoading

  useEffect(() => {
    activeSort && setUrlParams('activeSort', activeSort)
    sortKey && setUrlParams('sortKey', sortKey)
    if (searchKey) {
      setUrlParams('searchKey', searchKey)
    }
  }, [activeSort, searchKey, sortKey])

  const sortTable = useCallback((sortKey: keyof IApplication, type: SortTypes) => {
    setUrlParams('activeSort', type)
    setUrlParams('sortKey', sortKey)

    sortType.current = type
    setSortValue(sortKey)
  }, [])

  const handleInputChange = useCallback((value: string) => {
    setUrlParams('searchKey', value)
    setSearchValue(value)
  }, [])

  return (
    <div className='container'>
      <h2 className='applications_header'>Applications</h2>
      <div className='input-wrapper'>
        <InputBox
          placeholder='search here'
          onValueChange={handleInputChange}
          disabled={isLoading}
          value={searchValue || formattedSearchKey}
        />
      </div>
      <Table
        tableHeader={
          <TableHeader columns={columns} disabled={isLoading || errorOccured} onSort={sortTable} />
        }
      >
        {showApplications &&
          filteredApplications.map((application: IApplication) => (
            <TableRow key={application.id} rowData={formattedRowData(application)} />
          ))}
      </Table>

      <div className='info-wrapper'>
        {noResult && <InfoBox message='No Results Found' />}
        {errorOccured && <InfoBox message='Error Occured! Please try again' variant='error' />}
        {isLoading && <InfoBox message='Data is being loaded ...' />}
      </div>
    </div>
  )
}

export default Applications
