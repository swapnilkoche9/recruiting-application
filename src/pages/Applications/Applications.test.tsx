import { fireEvent, render, screen, within } from '@testing-library/react'

import { setupServer } from 'msw/node'
import { QueryClient, QueryClientProvider } from 'react-query'
import { Route, Routes, MemoryRouter } from 'react-router-dom'

import Applications from '.'
import { ApplicationsScenarios, baseHandlers, handlerForScenario } from './mocks/applicationsMock'

const server = setupServer(...baseHandlers())

beforeAll(() => {
  server.listen()
})
afterEach(() => server.resetHandlers())
afterAll(() => {
  server.close()
})

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      // âœ… turns retries off
      retry: false,
    },
  },
})

function setup() {
  return render(
    <MemoryRouter>
      <QueryClientProvider client={queryClient}>
        <Routes>
          <Route path='/' element={<Applications />} />
        </Routes>
      </QueryClientProvider>
    </MemoryRouter>,
  )
}

describe('Applications Page', () => {
  describe('Error flow', () => {
    it('should show error message  ', async () => {
      server.use(handlerForScenario[ApplicationsScenarios.ERROR]())
      setup()
      await screen.findByText('Error Occured! Please try again')
    })
  })
  describe('happy flow', () => {
    it('should render page ', async () => {
      const { container } = setup()
      await screen.findByText('Applications')

      expect(container).toMatchSnapshot()
    })

    it('should render all the applications ', async () => {
      setup()
      const applicationsList = await screen.findAllByTestId('table-row')

      expect(applicationsList.length).toBe(2)
    })

    it('sort ascending  ', async () => {
      setup()
      const ascArrowYoe = await screen.findByTestId('year_of_experience-asc')
      const ascArrowPa = await screen.findByTestId('position_applied-asc')
      const ascArrowAd = await screen.findByTestId('application_date-asc')
      fireEvent.click(ascArrowYoe)
      const applicationsList = await screen.findAllByTestId('table-row')
      within(applicationsList[0]).getByText('Alvin Satterfield')
      within(applicationsList[1]).getByText('Colette Morar')

      fireEvent.click(ascArrowPa)
      const applicationsList1 = await screen.findAllByTestId('table-row')
      within(applicationsList1[0]).getByText('Alvin Satterfield')
      within(applicationsList1[1]).getByText('Colette Morar')

      fireEvent.click(ascArrowAd)
      const applicationsList2 = await screen.findAllByTestId('table-row')
      within(applicationsList2[0]).getByText('Alvin Satterfield')
      within(applicationsList2[1]).getByText('Colette Morar')
    })

    it('sort descending  ', async () => {
      setup()
      const desArrowYoe = await screen.findByTestId('year_of_experience-des')
      const desArrowPa = await screen.findByTestId('position_applied-des')
      const desArrowAd = await screen.findByTestId('application_date-des')
      fireEvent.click(desArrowYoe)
      const applicationsList = await screen.findAllByTestId('table-row')
      within(applicationsList[1]).getByText('Colette Morar')
      within(applicationsList[0]).getByText('Alvin Satterfield')

      fireEvent.click(desArrowPa)
      const applicationsList1 = await screen.findAllByTestId('table-row')
      within(applicationsList1[1]).getByText('Colette Morar')
      within(applicationsList1[0]).getByText('Alvin Satterfield')

      fireEvent.click(desArrowAd)
      const applicationsList2 = await screen.findAllByTestId('table-row')
      within(applicationsList2[1]).getByText('Colette Morar')
      within(applicationsList2[0]).getByText('Alvin Satterfield')
    })
  })
})
