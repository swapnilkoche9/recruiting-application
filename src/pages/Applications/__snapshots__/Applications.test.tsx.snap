// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applications Page happy flow should render page  1`] = `
<div>
  <div
    class="container"
  >
    <h2
      class="applications_header"
    >
      Applications
    </h2>
    <div
      class="input-wrapper"
    >
      <input
        class="input"
        placeholder="search here"
        type="search"
        value=""
      />
    </div>
    <table
      class="table-container"
    >
      <thead
        class="table-header"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Email
          </th>
          <th>
            Age
          </th>
          <th>
            Years of experience
            <span
              class="arrow-wrapper"
            >
              <button
                data-testid="year_of_experience-asc"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 330 330"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m325.606 229.393-150.004-150a14.997 14.997 0 0 0-21.213.001l-149.996 150c-5.858 5.858-5.858 15.355 0 21.213 5.857 5.857 15.355 5.858 21.213 0l139.39-139.393 139.397 139.393A14.953 14.953 0 0 0 315 255a14.95 14.95 0 0 0 10.607-4.394c5.857-5.858 5.857-15.355-.001-21.213z"
                  />
                </svg>
              </button>
              <button
                data-testid="year_of_experience-des"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 25.93 25.93"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M25.397 4.554h-2.042l-9.974 12.644a.54.54 0 0 1-.835 0L2.575 4.554H.532a.533.533 0 0 0-.419.86l12.435 15.762c.104.125.255.2.419.2a.54.54 0 0 0 .416-.2L25.816 5.413a.524.524 0 0 0 .058-.561.524.524 0 0 0-.477-.298z"
                  />
                </svg>
              </button>
            </span>
          </th>
          <th>
            Position applied
            <span
              class="arrow-wrapper"
            >
              <button
                data-testid="position_applied-asc"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 330 330"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m325.606 229.393-150.004-150a14.997 14.997 0 0 0-21.213.001l-149.996 150c-5.858 5.858-5.858 15.355 0 21.213 5.857 5.857 15.355 5.858 21.213 0l139.39-139.393 139.397 139.393A14.953 14.953 0 0 0 315 255a14.95 14.95 0 0 0 10.607-4.394c5.857-5.858 5.857-15.355-.001-21.213z"
                  />
                </svg>
              </button>
              <button
                data-testid="position_applied-des"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 25.93 25.93"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M25.397 4.554h-2.042l-9.974 12.644a.54.54 0 0 1-.835 0L2.575 4.554H.532a.533.533 0 0 0-.419.86l12.435 15.762c.104.125.255.2.419.2a.54.54 0 0 0 .416-.2L25.816 5.413a.524.524 0 0 0 .058-.561.524.524 0 0 0-.477-.298z"
                  />
                </svg>
              </button>
            </span>
          </th>
          <th>
            Applied
            <span
              class="arrow-wrapper"
            >
              <button
                data-testid="application_date-asc"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 330 330"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m325.606 229.393-150.004-150a14.997 14.997 0 0 0-21.213.001l-149.996 150c-5.858 5.858-5.858 15.355 0 21.213 5.857 5.857 15.355 5.858 21.213 0l139.39-139.393 139.397 139.393A14.953 14.953 0 0 0 315 255a14.95 14.95 0 0 0 10.607-4.394c5.857-5.858 5.857-15.355-.001-21.213z"
                  />
                </svg>
              </button>
              <button
                data-testid="application_date-des"
                disabled=""
              >
                <svg
                  height="15px"
                  viewBox="0 0 25.93 25.93"
                  width="15px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M25.397 4.554h-2.042l-9.974 12.644a.54.54 0 0 1-.835 0L2.575 4.554H.532a.533.533 0 0 0-.419.86l12.435 15.762c.104.125.255.2.419.2a.54.54 0 0 0 .416-.2L25.816 5.413a.524.524 0 0 0 .058-.561.524.524 0 0 0-.477-.298z"
                  />
                </svg>
              </button>
            </span>
          </th>
          <th>
            Status
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      />
    </table>
    <div
      class="info-wrapper"
    >
      <h5
        class="error"
      >
        Error Occured! Please try again
      </h5>
    </div>
  </div>
</div>
`;
